src/java.base/share/classes/jdk/internal/foreign/MemorySegmentRenderUtil.java

```diff
@@ -0,0 +1,367 @@
+            return;
+        }
+        if (memoryLayout instanceof PaddingLayout paddingLayout) {
+            action.accept(state.indentSpaces() + paddingLayout.bitSize() + " padding bits");
+            state.indexAndAdd(paddingLayout);
+            return;
+        }
+        if (memoryLayout instanceof GroupLayout groupLayout) {
+
+            // Strictly, we should provide all permutations of nested unions.
+
+            final var separator = groupLayout.isStruct()
+                    ? ","  // Struct separator
+                    : "|"; // Union separator
+
+            action.accept(indentedLabel(state, groupLayout) + " {");
+            state.incrementIndent();
+            final var members = groupLayout.memberLayouts();
+            final long initialIndex = state.index();
+            long maxIndex = initialIndex;
+            for (int i = 0; i < members.size(); i++) {
+                if (groupLayout.isUnion()) {
+                    // If it is a union, we need to reset the index for each member
+                    state.index(initialIndex);
+                    // We record the max index used for any union member so we can leave off from there
+                    maxIndex = Math.max(maxIndex, state.index());
+                }
+                renderView(memorySegment, members.get(i), action, state, (i != (members.size() - 1)) ? separator : "");
+                if (groupLayout.isUnion()) {
+                    // This is the best we can do.
+                    state.index(maxIndex);
+                }
+            }
+            state.decrementIndent();
+            action.accept(state.indentSpaces() + "}" + suffix);
+            return;
+        }
+        if (memoryLayout instanceof SequenceLayout sequenceLayout) {
+            action.accept(indentedLabel(state, sequenceLayout) + " [");
+            state.incrementIndent();
+            final long elementCount = sequenceLayout.elementCount();
+            for (long i = 0; i < elementCount; i++) {
+                renderView(memorySegment, sequenceLayout.elementLayout(), action, state, (i != (elementCount - 1L)) ? "," : "");
+            }
+            state.decrementIndent();
+            action.accept(state.indentSpaces() + "]" + suffix);
+            return;
+        }
+        action.accept("Unknown memoryLayout: " + memoryLayout);
+    }
+
+    static String renderValueLayout(ViewState state,
+                                    ValueLayout layout,
+                                    String value,
+                                    String suffix) {
+        return indentedLabel(state, layout) + "=" + value + suffix;
+    }
+
+    static String indentedLabel(ViewState state,
+                                MemoryLayout layout) {
+        return state.indentSpaces() + layout.name()
+                .orElseGet(layout::toString);
+    }
+
+    static final class HexStreamState {
```

Comment:
While this class is simple - I note that there are many others hex utils in the JDK like src/java.base/share/classes/sun/security/util/HexDumpEncoder.java

Reply:
True. This class operates on input/output streams and so there is an impedance mismatch. But it would be possible to use it via an adaptor.